MAVLink for EdgeTx

olliw42 opened the pull request:
ok, so here we go, MAVLink for OpenTx for EdgeTx ...

this is working on my T16 for as much as I could test it

it is still missing some few important parts, namely the key and event handling and some lua drawing functions, and therefore there is not yet a usable lua script, like my script for opentx. That's so because the whole EdgeTx is not yet in a really usable state, and drawing and events handling has changed quite a bit. Risto's adaption of yappu's script might however work out of the box, if yappu's script works with EdgeTx.

However, the PR has ALL the framework. It should be way good enough to evaluate the suitability, and if merged before any YAML, RTOS, HAL or whatever similar massive changes, it could flow with it.

Concerning "closing doors" I should say that quite some pieces in the code are written such that they produce least code changes and essentially no merge conflicts. I obviously did so in order to keep my burden with upstream firmware changes low. However, this in reverse means that quite some code pieces are not such as one would write them or solve a thing if one would really integrate it well into the code. That is, if one would improve the code to make it fit well, which one eventually really should want to do, it would not merge seamlessly at all. An example of what I'm trying to say is e.g. the way how I've realized the external module bay handling. It works and does the job, but is anything than nice and well integrated into the whole fabric of the code, some better abstraction and architecture is desired.

Lastly, you'll find all my changes marked with //OW comments. These obviously will have to go away and I will certainly remove them once an OK for a merge is given. I find them however very usefull to work on the code and that's why they are there. As said, clearly, will go.

Have fun.

raphaelcoeffic commented on radio/src/dataconstants.h:
```diff
@@ -259,6 +259,11 @@ enum UartModes {
   UART_MODE_TELEMETRY,
   UART_MODE_SBUS_TRAINER,
   UART_MODE_LUA,
+//OW
```
Could you please remove these?

olliw42 replied:
when you give your ok I'll sure will clean up

raphaelcoeffic replied:
If I asked you to come up with a PR, that means that's where we are heading ;-)

raphaelcoeffic commented on radio/src/datastructs.h:
```diff
@@ -643,6 +643,20 @@ PACK(struct ModelData {
   char modelRegistrationID[PXX2_LEN_REGISTRATION_ID];
 
+//OW
+#if defined(TELEMETRY_MAVLINK)
```
I hope this is not too much to ask, but could you please make MAVLink activation a run-time option rather than a compilation option?

olliw42 replied:
I hear your words, but I'm frankly not sure how to do that without just spilling ram for everyone ... it would be not easy to do IMHO and probably would need major redesign of the memory usage and allocation
not sure you have seen that, but mavlink support is quite RAM hungry, and I hence had to bite into the .ram section

raphaelcoeffic replied:
After discussing it again with @rotorman, I think you're right. Your PR is quite disruptive in many aspects, so that it is probably safer to keep it as a build option for now. So, forget about that comment ;-)

raphaelcoeffic commented on radio/src/main.cpp:
```diff
@@ -30,7 +30,14 @@ uint8_t currentBacklightBright = 0;
 
+//OW
+#if defined(TELEMETRY_MAVLINK_USB_SERIAL)
+const char STR_USB_MAVLINK[] = "USB Mavlink (VCP)";
```
Could you please explain why this is necessary?

olliw42 replied:
connecting to usb with mavlink support, i.e., you can connect e.g. your PC or smarphone to the usb and talk mavlink, very handy

raphaelcoeffic replied:
That is understood, but would the normal USB serial be ok? Or is it necessary to have a particular string here?

olliw42 replied:
you somehow need a way to select that it should be linked with the mavlink router, which it normally would not be ... so, some setting somewhere is needed to distinguish ... doing it that way is just super convenient because you plug in the usb and select as you do anyhow for joystick or vcp 
TANSTAAFL

raphaelcoeffic replied:
I get that point, no problem with that! It's just, why needing a different name here? Or is that creating a second VCP (on top of the debug one, in case compiled with debug options). You could probably just activate VCP, and that's it. Support for it is already there, as you can see.

olliw42 replied:
???
I'm not sure you got the point: If you have two options among which to select and both are named the very same how would you then ever know which one you select ???

* usb
* usb

so, now please select the one which links you to mavlink

> Support for it is already there, as you can see.
yes, and I'm obviously using it, don't I

I'm frankly totally lost

rotorman replied:
@olliw42 It's not usb & usb, but Joystick and Mass Storage at the moment. What the MAVLink extensions need, is virtual serial port (VCP, or officially CDC). What @raphaelcoeffic says is that you already have a VCP there, e.g. if you build with DEBUG and CLI options set (vanilla OTX/ETX) and then hook the top USB-C to PC, you have three options pop-up. The serial option, that you need for MAVLink Routing via top USB-C is already there, so no need to redefine it with own name. I hope I was able to convey the meaning and clear up the issue a bit. Best!

olliw42 replied:
come on guys
we are talking about 
* VCP for serial
* VCP for serial with mavlink connection

so, make it 

* vcp
* vcp

if you want but the argument does not change at all

of course there can be situations, depending on flags, there only one of them might be available, it might even not be available at all LOL, but really, you want to change names depending on that, so that users need to know what their firmware ??? One could do as you suggest, all I'm saying is that it's IMHO short sighted.

raphaelcoeffic replied:
I don't want to change names at all! this it it ;-) These names are purely informational and not required for the function.

raphaelcoeffic commented on radio/src/mavlink_telem_camera.cpp:
```diff
@@ -0,0 +1,268 @@
+
+#include "opentx.h"
+
+constexpr float FPI = 3.141592653589793f;
```
I’d suggest moving the complete telemetry code into the corresponding directory.

olliw42 replied:
I was thinking about that ... but refrained form moving it to telemetry, since - despite the saying - mavlink is actually not telemetry and really functions totally differently and hence has also very different resource usage and code layout ... i.e. architecturally it doesn't fit into that folder and I have difficulties considering it as the "corresponding directory" :)
I was thinking about a folder mavlink, but it may confuse with the thirdparty folder, or a folder mavlink_telem...
Maybe a folder /telemetry/mavlink_telem ??

raphaelcoeffic replied:
A mavlink folder inside telemetry is perfect for me.

olliw42 replied:
ok

olliw42 replied:
done

raphaelcoeffic replied:
Thx!

raphaelcoeffic commented on radio/src/opentx.cpp:
```diff
@@ -225,6 +225,13 @@ void per10ms()
 
+//OW
+#if defined(TELEMETRY_MAVLINK)
+  mavlinkTelem.tick10ms();
```
We need to find a way to include this into the normal telemetry processing path, if that is possible. And please try to make it conditional on MAVLink serial being activated. Does it make sense?

olliw42 replied:
I don't disagree with either, but both IMHO are not so easy ... 
for 1. see my response below to https://github.com/EdgeTX/edgetx/pull/26#discussion_r633081750
for 2. see what I've said in the above

raphaelcoeffic replied:
Ok, we will review this again once I get into the details.

raphaelcoeffic commented on radio/src/opentx.h:
```diff
@@ -27,6 +27,66 @@
 #include "touch.h"
 
+//OW
+#define OWVERSIONSTR  "v27"
```
Please a more “neutral” name.

olliw42 replied:
when you give your ok I'll sure will clean up

olliw42 replied:
done

raphaelcoeffic replied:
Thx!

raphaelcoeffic commented on radio/src/targets/common/arm/stm32/usb_driver.cpp:
```diff
@@ -79,6 +79,11 @@ void usbStart()
     case USB_SERIAL_MODE:
+//OW
+#if defined(TELEMETRY_MAVLINK)
+    case USB_MAVLINK_MODE:
```
This is most probably not necessary, but I might be wrong.

olliw42 replied:
it won't hurt, so it's not necessary

olliw42 replied:
done

raphaelcoeffic commented on radio/src/targets/common/arm/stm32/usb_driver.h:
```diff
@@ -30,6 +30,11 @@ enum usbMode {
 #if defined(USB_SERIAL)
+//OW
+#if defined(TELEMETRY_MAVLINK)
+  USB_MAVLINK_MODE,
```
Same here

olliw42 replied:
dito

olliw42 replied:
done

raphaelcoeffic commented on radio/src/targets/horus/board.h:
```diff
@@ -153,10 +153,18 @@ void SDRAM_Init();
-#define EXTERNAL_MODULE_ON()           GPIO_SetBits(EXTMODULE_PWR_GPIO, EXTMODULE_PWR_GPIO_PIN)
-#define EXTERNAL_MODULE_OFF()          GPIO_ResetBits(EXTMODULE_PWR_GPIO, EXTMODULE_PWR_GPIO_PIN)
+//OW
+//#define EXTERNAL_MODULE_ON()           GPIO_SetBits(EXTMODULE_PWR_GPIO, EXTMODULE_PWR_GPIO_PIN)
```
Interesting: these defs are not used normally?

olliw42 replied:
they are used, that's why I repurpose them

raphaelcoeffic replied:
Sorry, I think I don't understand the "re-purposing" :-( Could you please elaborate on that?

olliw42 replied:
well, these macros are used throughout the code, but in their original form they don't do what I need them to do ... so I either could have reworked all the code or simply replace them and make them call what is needed ... quite standard approach and in fact used lots and lots in the opentx/edgetx code ...

raphaelcoeffic commented on radio/src/targets/horus/hal.h:
```diff
@@ -477,6 +477,49 @@
 #endif
 
+//OW
+#if (defined(RADIO_T16) || defined(RADIO_T18)) && defined(AUX_SERIAL)
```
Which serial port is that?

olliw42 replied:
AUX = uart3, AUX2 = BT = GPS = uart6

raphaelcoeffic replied:
Thx

raphaelcoeffic commented on radio/src/targets/horus/telemetry_driver.cpp:
```diff
@@ -28,6 +28,11 @@ DMAFifo<TELEMETRY_FIFO_SIZE> telemetryDMAFifo __DMA (TELEMETRY_DMA_Stream_RX);
 #endif
 
+//OW
+extern Fifo<uint8_t, 32> mavlinkTelemExternalTxFifo_frame;
```
Is there a possibility the same buffers could be used as other telemetry protocols, or is that necessary on top of the others?

olliw42 replied:
in principle it is possible to get away with less buffers
but that's btw so also for the telemetry protocols themselves, it's quite a mess frankly
what it actually would need is a pretty complete overhaul of the whole protocol and and module and serial handling. IMHO.

raphaelcoeffic replied:
I don't expect you to use smaller buffers! I just thought we could have the buffer size being some "#define", and have their size being cranked up when compiled with MAVlink support. Does that make sense?

olliw42 replied:
I'm somewhat lost here too ... my answer would be just a rewording of my previous answer, not sure what else I should say

this can be done in 10000 ways, and 9000 of them are equally good ...

I'm not a friend of what you suggest because this just leads to exactly this miriad of tiny flags doing tiny things ... I rather would go with what I suggested, but this shouldn't be part of this PR
it's frankly also somewhat ironical since you here suggest to add flags for doing tiny things yet constantly complain about each flag that they should by run time ... it's still a mistery to me what architecture you actually strive for ... if there is such a plan

raphaelcoeffic commented on radio/src/tasks.cpp:
```diff
@@ -41,6 +41,11 @@ void stackPaint()
   cliStack.paint();
 #endif
+//OW
+#if defined(TELEMETRY_MAVLINK)
```
Oho! You added a new task?

olliw42 replied:
yeah, the stuff needs to be done somewhere, and the mixer task is too frequent and too tight already, and the other tasks way too low rate

raphaelcoeffic replied:
Ok, understood!

olliw42 commented:
so, new update:
* moved mavlink files to telemetry/mavlink folder
* removed the #ifdef wrappers around USB_MAVLINK_MODE
* cleanified the OWVERSIONSTR
* I've slightly change a thing with the usb vcp

I have not removed the //OW markers all around ... I just again realized how much I need them to work on the code additions to find my changes ... but as said before, I will ensure that they will all go before merge (but it will be the last thing I will do)

the other points raised don't seem controversial to me but settled

raphaelcoeffic commented:
> the other points raised don't seem controversial to me but settled

Right, I need to go into deeper inspection now. That was only the first superficial look. Let's make this perfect!

olliw42 commented:
> Let's make this perfect!

impossible at this point in time 

you can't make something perfect if it's embedeed in a non-perfect environment ;)

I'm going to close this here since I screwed up my fork and lost the ability to push to this branch.
